discovery_engine_core/core/src/engine.rs

```diff
@@ -1109,18 +1118,236 @@ impl Engine {
         let exclusion_set = excluded.iter().cloned().collect::<HashSet<_>>();
         *self.endpoint_config.excluded_sources.write().await = excluded;
 
-        let mut stacks = self.stacks.write().await;
-        for stack in stacks.values_mut() {
-            stack.prune_by_sources(&exclusion_set, true);
-        }
-        drop(stacks); // guard
-        self.exploration_stack
-            .prune_by_sources(&exclusion_set, true);
-
+        self.filter_excluded_sources_for_all_stacks(&exclusion_set)
+            .await;
         self.update_stacks_for_all_markets(history, sources, self.core_config.request_new)
             .await
     }
 
+    /// Sets a new list of excluded and trusted sources.
+    #[cfg_attr(not(feature = "storage"), allow(unused_variables))]
+    pub async fn set_sources(
+        &mut self,
+        sources: &[WeightedSource],
+        excluded: Vec<String>,
+        trusted: Vec<String>,
+    ) -> Result<(), Error> {
+        #[cfg(feature = "storage")]
+        {
+            let trusted_set = excluded.iter().cloned().collect::<HashSet<_>>();
```

Comment:
```suggestion
            let trusted_set = trusted.iter().cloned().collect::<HashSet<_>>();
```

Reply:
Oh boy! I checked it several times but I missed this one 
Nice catch!